[Multilane] Adds continuity constraints into the loader

agalbachicar opened the pull request:
This PR goes on top of #8302.

In this PR you'll find:

- `EndpointZ::theta_dot`'s type is changed from `double` to `drake::optional<double>`.
- Forces a continuity constraint to match two roads connecting road curves.
- Modifies the loader to identify if a start or end point belongs to either a reference curve or a lane curve.
- Modifies loader tests to provide code coverage to the continuity constraint involved.

A follow up PR will provide Builder support to build lane-connected segments.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8676)
<!-- Reviewable:end -->

agalbachicar commented:
+@maddog-tri for feature review please.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LAy7uVQ-2ufHONoAI8n:b-5klkw7)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I just updated this PR with recent changes from master. Code should not be affected.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LD3IEKK3UvsNAhZJZTa:bdsst71)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
@drake-jenkins-bot linux-xenial-gcc-cmake-experimental-matlab please

maddog-tri commented:
Reviewed 4 of 10 files at r1.
Review status: 4 of 10 files reviewed at latest revision, all discussions resolved.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDOLixT9Oxb8HneA3dE:-LDOLixT9Oxb8HneA3dF:b-ma429x) (no related file):*
Whoa, hang on a minute.... I dove into reviewing this and was cranking along and just now realized the premise is wrong.

Continuity constraints don't belong in the loader.  They belong in the _Builder_.

It's the Builder's job to make sure things link up continuously.  Conversely, one should be able to hook together a smooth road network using just the Builder --- that's what that whole fluent API is all about.

The Loader should be really simple.  It shouldn't have any math in it.  All it needs to do is translate the YAML directly into Builder API calls.  Likewise, all the Loader's unit-tests need to do is to check that translation.  The Loader doesn't need to know anything about continuity.

---

*[automotive/maliput/multilane/connection.h, line 70 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDIzu3Y7Yw5Ukqu2wtt:-LDIzu3Y7Yw5Ukqu2wtu:b-zgm4k3) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/connection.h#L70)):*
> ```objc
> ///           when theta > 0, elevation at r > 0 is above elevation at r < 0)
> ///  - theta_dot: rate of change of superelevation with respect to arc length
> ///               of the reference path
> ```

Should document that theta_dot is optional and why.

---

*[automotive/maliput/multilane/connection.h, line 81 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDJ-1lB-_vUvcZKyRdE:-LDJ-1lB-_vUvcZKyRdF:bw7z9c9) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/connection.h#L81)):*
> ```objc
> 
>   EndpointZ(double z, double z_dot, double theta, double theta_dot)
>       : z_(z), z_dot_(z_dot), theta_(theta), theta_dot_(theta_dot) {}
> ```

BTW:  It could be more clear/simple to combine these into a single constructor with the `theta_dot` parameter typed as an `optional`.

---

*[automotive/maliput/multilane/connection.h, line 90 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDIzD0R4C0mY0LtkSdF:-LDIzD0R4C0mY0LtkSdG:beu8xa0) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/connection.h#L90)):*
> ```objc
>   EndpointZ reverse() const {
>     DRAKE_DEMAND(theta_dot_.has_value());
>     return EndpointZ(z_, -z_dot_, -theta_, theta_dot_.value());
> ```

Does `theta_dot_` really need to have a value?

The `optional` could simply be copied (especially since reverse() would not affect its value).

---

*[automotive/maliput/multilane/loader.cc, line 65 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDJIUyV3AdtFl4SowdW:-LDJIUyV3AdtFl4SowdX:b1w0cc5) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L65)):*
> ```cpp
> // second item will be y coordinate and the third item will be heading angle in
> // degrees.
> EndpointXy endpointxy(const YAML::Node& node) {
> ```

(META) Channeling platform reviewers:  it occurs to me that none of these "parse from YAML" functions qualifies for a lower_case_only name.  Maybe rename them to `ParseEndpointXy()`, etc?

---

*[automotive/maliput/multilane/loader.cc, line 74 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDJJZ_FBdhnW_po5YUd:-LDJJZ_FBdhnW_po5YUe:bx9ncvm) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L74)):*
> ```cpp
> // Parses a YAML `node` and returns an EndpointZ object from it.
> // `node` must be a sequence of three or four doubles, the last item is not
> // mandatory. The first two items are z coordinate and z_dot. The last
> ```

BTW:  Why is "is not mandatory" preferable to "is optional"?

---

*[automotive/maliput/multilane/loader.cc, line 75 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDJJRzzCKJZMCPQPBMg:-LDJJRzzCKJZMCPQPBMh:b-wnjk2a) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L75)):*
> ```cpp
> // `node` must be a sequence of three or four doubles, the last item is not
> // mandatory. The first two items are z coordinate and z_dot. The last
> // two items are theta angle and theta_dot which are expressed in degrees and,
> ```

This comma looks misplaced.

---

*[automotive/maliput/multilane/loader.cc, line 77 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDJK25d-OtXO3-NBeLn:-LDJK25d-OtXO3-NBeLo:b-vq1l0x) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L77)):*
> ```cpp
> // two items are theta angle and theta_dot which are expressed in degrees and,
> // degrees per meter respectively.
> // theta_dot is optional since continuity constraints may need to be adjusted.
> ```

FYI:  suggestion for that last sentence: "theta_dot is optional since, in general, its value must be derived from other endpoint/connection parameters in order to preserve continutiy."

---

*[automotive/maliput/multilane/loader.cc, line 86 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDJL1C8B5nDDYtAp_o-:-LDJL1C90eMtax3fJ8dH:b-av31gy) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L86)):*
> ```cpp
>                          deg_to_rad(node[3].as<double>()))
>              : EndpointZ(node[0].as<double>(), node[1].as<double>(),
>                          deg_to_rad(node[2].as<double>()));
> ```

BTW: (See... if there was one constructor that took an `optional<double>` for the last parameter, you could push the conditional into the last parameter only.)

---

*[automotive/maliput/multilane/loader.cc, line 250 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDOGeui6-116R-oqEUg:-LDOGeuj0Wm5vQPEY9y8:b9i6h76) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L250)):*
> ```cpp
> // elevation with respect to arc length of the reference path) and the
> // curvature at one end point of connection's curve A. And let theta_dot_B,
> // z_dot_B, K_B be the same magnitudes at one end point of connection's curve
> ```

"magnitudes" --> "values"  ("magnitudes" implies non-negative.)

---

*[automotive/maliput/multilane/loader.cc, line 262 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDNkhQE1IzxVgMb5GLY:-LDNkhQE1IzxVgMb5GLZ:bazuvyi) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L262)):*
> ```cpp
> // theta_dot = K * sin(-atan(z_dot))
> void ComputeContinuityConstraint(double curvature, EndpointZ* endpointz) {
>   DRAKE_DEMAND(curvature >= 0.);
> ```

`curvature` is a signed quantity (sign indicates to-left/to-right).

Implies:  unit-tests need to exercise negative curvature.

---

*[automotive/maliput/multilane/loader.cc, line 274 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDNjgbdCitERWrtyVu-:-LDNjgbdCitERWrtyVu0:b-95g7qz) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L274)):*
> ```cpp
> // theta_dot is nullopt.
> // When `type` is either KReference or kLane, the continuity constraint needs
> // to be computed.
> ```

It's not clear from this docstring that this method is focused on checking and/or computing the value of `theta_dot`.

---

*[automotive/maliput/multilane/loader.cc, line 275 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDNjMciD_sfu2lYTArH:-LDNjMciD_sfu2lYTArI:becbn25) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L275)):*
> ```cpp
> // When `type` is either KReference or kLane, the continuity constraint needs
> // to be computed.
> void EvaluateContinuityConstraint(EndpointType type, double curvature,
> ```

Perhaps `EnforceContinuityConstraint()` is a better name?

---

*[automotive/maliput/multilane/loader.cc, line 277 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDNkN155oO9IFQ8WJ0l:-LDNkN155oO9IFQ8WJ0m:bazuvyi) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L277)):*
> ```cpp
> void EvaluateContinuityConstraint(EndpointType type, double curvature,
>                                   EndpointZ* ez_point) {
>   DRAKE_DEMAND(curvature >= 0.);
> ```

`curvature` is a signed quantity (sign indicates to-left/to-right).

Implies:  unit-tests need to exercise negative curvature.

---

*[automotive/maliput/multilane/loader.cc, line 294 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDODV4z9t2z9L8Y2waF:-LDODV4z9t2z9L8Y2waG:b8ue86y) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L294)):*
> ```cpp
>       ComputeContinuityConstraint(curvature, ez_point);
>       break;
>     }
> ```

No need to repeat the identical blocks for `kReference` and `kLane`; just stack the `case`s.

---

*[automotive/maliput/multilane/loader.cc, line 296 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDOE4Si8-yWIN61emPS:-LDOE4Si8-yWIN61emPT:bab7kxp) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L296)):*
> ```cpp
>     }
>     default: { DRAKE_ABORT(); }
>   }
> ```

Just to ensure that I understand the logic here (and maybe some more comments would help), is the idea that:
  * If the source (`type`) of the endpoint is `kReference` or `kLane`, then (1) it is safe to assume the user wants the new connection to be continuous to the source connection and (2) `theta_dot` will already be populated with a value that was appropriate for the source connection.  So, it is reasonable to always stomp that `theta_dot` and replace it with a value that ensures continuity.
  * If the source (`type`) is `kRaw`, then either the user has a specific `theta_dot` value in mind (and the value is non-empty) or the user is interested in continuity (and the value is empty).  So, we don't replace a pre-existing value.
Yes?

Maybe it would be good to `DRAKE_DEMAND(theta_dot().has_value())` in the kReference/kLane case, since that is what is always expected.

---

*[automotive/maliput/multilane/loader.cc, line 340 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDOHhVmDXF9xxmdxj1T:-LDOHhVmDXF9xxmdxj1U:bdbxt1v) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L340)):*
> ```cpp
>   } else {
>     arc_offset = ResolveArcOffset(node["arc"]);
>     curvature = 1. / arc_offset.radius();
> ```

This `curvature` needs to be signed (e.g., get the correct sign/polarity from `d_theta_`).

---

*[automotive/maliput/multilane/loader.cc, line 444 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDNfqCqEUE_n-1mpe4D:-LDNfqCqEUE_n-1mpe4E:bvqjd12) ([raw file](https://github.com/robotlocomotion/drake/blob/93bec79f9ab01cbf9be9e858c6fa805f08b2d4a0/automotive/maliput/multilane/loader.cc#L444)):*
> ```cpp
> 
>   for (int i = 0; i < conn.num_lanes(); ++i) {
>     curvature = conn.LaneCurvature(i);
> ```

Something about using the "lane curvature", instead of always doing things w.r.t. the actual reference-curve curvature, seems troublesome....

(Reminder to me to work through the math)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDOLixT9Oxb8HneA3dE:-LDi3hrpC5CAr7CbR9Tn:bshrxj7) (no related file):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

Yes, you're right. Although, I see an issue with the use of the fluent API as it is. We can't ask the Builder to adjust theta_dot when the full `EndpointZ` (`at(endpoint, direction)` or `z_at(endpoint_z, direction)` methods) is given because there is no way to tell whether that `Spec` was built from a reference-curve / lane-curve or a raw `Endpoint` and we need or we don't need theta_dot adjustment.
</blockquote></details>

I'd think that the `theta_dot` value in a `Spec` should never be set unless the caller wants a specific value.  Otherwise, `theta_dot` should be left unset, and only then will the Builder calculate the appropriate value. The Builder should not change any theta_dot values that have already been set.

In general, YAML authors would leave the theta-dot parameter unset (including for raw points), and the Loader will propagate that unset value to the Builder.

That works, right?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDOLixT9Oxb8HneA3dE:-LDmunee9hdqrFwiTP0y:bsasxis) (no related file):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

So, to put this in code:

```
builder.Connect(
    "connection_name",
    lane_layout,
    StartReference().at(other_connection, Which::kEnd, Direction::kForward),
    LineOffset(100.),
    EndReference().z_at(z_end, Direction::kReverse));
```

Here, `StartReference::at()` will query the `other_connection` to get the end endpoint whose theta_dot is not cleared by `other_connection`. In this case, inside `at()`, theta_dot is cleared and then the `builder` will adjust it.

It's not the same case for `EndReference::z_at()` where `z_end`'s theta_dot will be respected if provided, otherwise `builder` will adjust it.

Are we on the same page?


</blockquote></details>

FYI, latest commit implements my last comment so it can easily be reverted if necessary.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDOLixT9Oxb8HneA3dE:-LDqdw7DCcr3GyF34No3:b-kgtjqh) (no related file):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

FYI, latest commit implements my last comment so it can easily be reverted if necessary.
</blockquote></details>

Thanks for the example --- yes, I think that captures the idea, and I think we are on the same page now.

Basically:  any use of `at(other_connection, ...)` or `z_at(other_connection, ...)` implies that the user wants continuity with that other connection, so the Builder selects theta_dot appropriately.  On the other hand, in the case of `at(literal_endpoint, ...)` or `z_at(literal_endpoint, ...)`, the Builder takes the user's literal_endpoint at face value, and thus the user either chooses a specific theta_dot, or leaves the decision to the Builder by leaving the value unspecified (which would generally be the case, unless the user is trying to make a corkscrew, etc).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
I've done another pass on the code-code (but still need to look at the test-code).  PTAL.

---

Review status: 7 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[automotive/maliput/multilane/connection.h, line 90 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDIzD0R4C0mY0LtkSdF:-LDi-sHF59SZd3TE91vL:b-56lgli) ([raw file](https://github.com/robotlocomotion/drake/blob/14fc75353c5310fdaf8d0f2f9dce64691a1ea5c0/automotive/maliput/multilane/connection.h#L90)):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

"`optional` could simply be copied":

Just `return EndpointZ(z_, -z_dot_, -theta_, theta_dot_);` (without any conditionals) should work.

---

*[automotive/maliput/multilane/connection.cc, line 202 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDw-kHcETFeZcyGyLbH:-LDw-kHcETFeZcyGyLbI:b-7h7p23) ([raw file](https://github.com/robotlocomotion/drake/blob/9181d4143eee3e043e2c94905b14d6068a7c5ca2/automotive/maliput/multilane/connection.cc#L202)):*
> ```cpp
>       radius_ - std::copysign(1., d_theta_) * lane_offset(lane_index);
>   return 1. / lane_radius;
> }
> ```

I think `ReferenceCurvature()` and `LaneCurvature()` can go away; I don't think they will be used/necessary anymore.  

(But if they are needed, maybe it is better if they provide a signed-curvature, and are documented accordingly.)

---

*[automotive/maliput/multilane/loader.cc, line 45 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDw4u1o7o9zn4yItmM0:-LDw4u1o7o9zn4yItmM1:b7f8d8l) ([raw file](https://github.com/robotlocomotion/drake/blob/9181d4143eee3e043e2c94905b14d6068a7c5ca2/automotive/maliput/multilane/loader.cc#L45)):*
> ```cpp
>   // connection's curve. Given that, a curvature can be assigned.
>   optional<double> curvature{};
> };
> ```

So... none of `EndpointType`, `CurveType`, `LoadedEndpoint` is necessary anymore, right?  Nor `EnforceContinuityConstraint()`, nor the catalogs of all reference and lane endpoints.

Since the loader just needs to forward parameters to the builder, I think the only thing the loader needs to keep track of is (1) a map of the endpoints which it has parsed from the `points:` in YAML, and (2) a map of the connections which it has had the builder construct, from the `connections:` in YAML.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LDw7JLC9JG-PVjJfHx5:bz39p5y)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks!

Comments were addressed. PTAL

---

Review status: 2 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[automotive/maliput/multilane/connection.h, line 90 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDIzD0R4C0mY0LtkSdF:-LDwmuAJ70c1WC03H3GP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/14fc75353c5310fdaf8d0f2f9dce64691a1ea5c0/automotive/maliput/multilane/connection.h#L90)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

"`optional` could simply be copied":

Just `return EndpointZ(z_, -z_dot_, -theta_, theta_dot_);` (without any conditionals) should work.
</blockquote></details>

Done.

---

*[automotive/maliput/multilane/connection.cc, line 202 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDw-kHcETFeZcyGyLbH:-LDwn7_e-6XnqJO5tNpr:b-69qvg2) ([raw file](https://github.com/robotlocomotion/drake/blob/9181d4143eee3e043e2c94905b14d6068a7c5ca2/automotive/maliput/multilane/connection.cc#L202)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

I think `ReferenceCurvature()` and `LaneCurvature()` can go away; I don't think they will be used/necessary anymore.  

(But if they are needed, maybe it is better if they provide a signed-curvature, and are documented accordingly.)
</blockquote></details>

Done.

Yes, I missed it. Thanks.

---

*[automotive/maliput/multilane/loader.cc, line 45 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LDw4u1o7o9zn4yItmM0:-LEEpDja3k2yyve0EadT:bfe6t0c) ([raw file](https://github.com/robotlocomotion/drake/blob/9181d4143eee3e043e2c94905b14d6068a7c5ca2/automotive/maliput/multilane/loader.cc#L45)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

So... none of `EndpointType`, `CurveType`, `LoadedEndpoint` is necessary anymore, right?  Nor `EnforceContinuityConstraint()`, nor the catalogs of all reference and lane endpoints.

Since the loader just needs to forward parameters to the builder, I think the only thing the loader needs to keep track of is (1) a map of the endpoints which it has parsed from the `points:` in YAML, and (2) a map of the connections which it has had the builder construct, from the `connections:` in YAML.
</blockquote></details>

Done.

Only `points:`and `connections:` are kept now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LEJsrRB117quzLiBA83:bt2kte6)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
*[automotive/maliput/multilane/loader.cc, line 39 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEojW20DWRn72NPp_7e:-LEp1Uzm7Ajyt9JI8wjm:b-clugj9) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L39)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

META:  Instead of doing all the repetitive `find()` and `rfind()` and `*_pos != npos` to parse/reparse endpoint references, how about using regular expressions?  One regex for `points...` and one for `connections...`.  Each reference type has a fixed number of period-separated fields, so the syntax for each should be pretty simple.

I think you could have one method that parses both point and connection references into components stored in a struct, and it would simplify the code in this file a lot.  E.g., jamming both reference types into a single struct:
```
struct ParsedReference {
  ReferenceType type;  \\ e.g., some enum kPoint vs kConnection
  std::string id;
  Direction direction;  \\ kForward vs kReverse
  optional<api::LaneEnd::Which> end;  \\ kStart vs kFinish
  optional<int> lane_id;  \\ with nullopt == reference curve
};
```
</blockquote></details>

I preferred this way because of bad experience with regex with C++ because of bad experience ( [ignition::rndf](https://bitbucket.org/ignitionrobotics/ign-rndf/src) parser ). It was really slow. I don't have a strong opinion, just providing some feedback. Do you still prefer to move forward with regular expressions?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
*[automotive/maliput/multilane/loader.cc, line 39 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEojW20DWRn72NPp_7e:-LEpASoH0pyljhrySnvY:bcoqgv9) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L39)):*
<details><summary><i>Previously, agalbachicar (Agustin Alba Chicar) wrote…</i></summary><blockquote>

I preferred this way because of bad experience with regex with C++ because of bad experience ( [ignition::rndf](https://bitbucket.org/ignitionrobotics/ign-rndf/src) parser ). It was really slow. I don't have a strong opinion, just providing some feedback. Do you still prefer to move forward with regular expressions?
</blockquote></details>

Really?  (How long ago was that?  I wonder if it has improved.  Regex is so useful.)  Even without using regex, though, I think consolidating all the reference string parsing in one place and at one time will make the code a lot more compact and clear (and efficient as well).  It's just `.` separated tokens.

(PS:  I recall that the syntax is set up to be permissive and allow periods within ID's, since the ID is the only token that is a free-form user-supplied string, and the number of tokens/periods for each reference type is fixed.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
*[automotive/maliput/multilane/loader.cc, line 39 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEojW20DWRn72NPp_7e:-LEpI3B48tlije2atcNw:b-swvsk1) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L39)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Really?  (How long ago was that?  I wonder if it has improved.  Regex is so useful.)  Even without using regex, though, I think consolidating all the reference string parsing in one place and at one time will make the code a lot more compact and clear (and efficient as well).  It's just `.` separated tokens.

(PS:  I recall that the syntax is set up to be permissive and allow periods within ID's, since the ID is the only token that is a free-form user-supplied string, and the number of tokens/periods for each reference type is fixed.)
</blockquote></details>

Yes, it used C++11 and a year and half ago...

OK, no problem, I'll go that way.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks Matt, comments were addressed. PTAL.

---

Review status: 7 unresolved discussions, LGTM missing from assignee maddog-tri, platform LGTM missing

---

*[automotive/maliput/multilane/connection.h, line 92 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEWSrxD6ZxMvTTt-tHh:-LEzR8s31DnULUZZuBCK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/connection.h#L92)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Commented out code.  (That conditional is fighting for its survival, any way it can....)
</blockquote></details>

Done.

---

*[automotive/maliput/multilane/connection.cc, line 41 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEWTg-J0XFS_c90OSyn:-LEuLKBlCyYmAcaDXFeO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/connection.cc#L41)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Don't you want this to be `line_offset`?

That's what you changed it to in an earlier revision, and what it still is in the .h file.  (I suspect it got half-reverted when rebasing on top of the last PR from Michel.)
</blockquote></details>

Done.

---

*[automotive/maliput/multilane/loader.cc, line 27 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEocHhe8NLmSRcjfi7G:-LEuKPDj-EgFgmCvtR9L:b-y5qtit) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L27)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Is the keyword `connection` or `connections`?  (Docstring says one here, kConstant below says the other.)
</blockquote></details>

Good catch, I missed it. Thanks.

---

*[automotive/maliput/multilane/loader.cc, line 39 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEojW20DWRn72NPp_7e:-LEzR5wE3DikgDivmhwC:b-njgs6u) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L39)):*
> ink you could have one method that parses both point and connection references into components stored in a struct, and it would simplify the code in this file a lot.  E.g., jamming both reference types into a single struct:
Done.

---

*[automotive/maliput/multilane/loader.cc, line 146 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEoeVC70IYqj7RDNQ9W:-LEuKLHf4c1ax4f390qe:bjaqb2p) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L146)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

BTW:  Typically, we use empty `{}` initializers to indicate "initialized with meaningless default value that is intended to be overwritten with something meaningful", which I think is the case here.  (The three zeroes are not meaningful; they are supposed to be overwritten by the result of ParseLanes().)
</blockquote></details>

Done. I missed it, thanks.

---

*[automotive/maliput/multilane/loader.cc, line 184 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEog8h20OwBcQISyIfc:-LEuDgxn6WbjJm89bved:bqfjivm) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L184)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

By "its curve" do you mean "its reference curve or one of its lanes"?
</blockquote></details>

Yes, I mean that.

Changed to your quoted string.

---

*[automotive/maliput/multilane/loader.cc, line 350 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEooT6h4MvYir9-uuXu:-LEuKF8XD2gQZTZQU1no:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L350)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

BTW:  Rename to `Parse...()`?  ("Resolve" typically involves "resolve a reference", which is the operation to look-up what a specific reference points to, not the operation which creates the index.)
</blockquote></details>

Done.

---

*[automotive/maliput/multilane/loader.cc, line 364 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEos7yWECpzmkMFZEVU:-LEuKDzWEMLDJDf0iRcq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/loader.cc#L364)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

BTW:  rename `xyz_catalog` to `point_catalog` for clarity?  (since code talks about "point references" versus "connection references")
</blockquote></details>

Done.

---

*[automotive/maliput/multilane/test/multilane_builder_test.cc, line 712 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEWhGOIACl0cXjlBapZ:-LEuKAhm4yqwmGOla7kZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/test/multilane_builder_test.cc#L712)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

How were these theta_dot values (here and below) calculated?

It would be helpful to show the math in a comment (or maybe even execute the expression instead of a long literal number).
</blockquote></details>

Done.

---

*[automotive/maliput/multilane/test/multilane_loader_test.cc, line 697 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LEp-hLFCQ8_hBd4aLyE:-LEuK8MeBMM-RzA4nNCB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aaa7ce5f8946f10df338f6905e5ccb39dfd6c35a/automotive/maliput/multilane/test/multilane_loader_test.cc#L697)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

BTW: Does `modified` mean `theta_dot is cleared`?  (Better to be explicit about what is supposed to happen.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LEzUBgf2a6FKB3oMThm:b-4txnkp)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, except for some complaints about the regular expressions.

---

Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[automotive/maliput/multilane/loader.cc, line 44 at r5](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF-LBVs34D_iqt2BlNp:-LF-LBVs34D_iqt2BlNq:bnt5j3v) ([raw file](https://github.com/robotlocomotion/drake/blob/ce1efe03236c3eff1df8b5b54c0ead8461b4e3b3/automotive/maliput/multilane/loader.cc#L44)):*
> ```cpp
>   optional<api::LaneEnd::Which> end;
>   // When `type` == kConnection, is `id` connection's lane index. Otherwise, it
>   // must be nullopt.
> ```

BTW: ", or nullopt if referring to a connection's reference curve." would be more clear about what nullopt means.

---

*[automotive/maliput/multilane/loader.cc, line 237 at r5](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF-Qkkm5cznUoveVvll:-LF-Qkkm5cznUoveVvlm:b-rcv9se) ([raw file](https://github.com/robotlocomotion/drake/blob/ce1efe03236c3eff1df8b5b54c0ead8461b4e3b3/automotive/maliput/multilane/loader.cc#L237)):*
> ```cpp
>       "^(connections.)((?:[\\w\\-_]+\\.)+)(start.|end.)((ref|\\d+)\\.)+"
>       "(forward|reverse)$",
>       std::regex::ECMAScript};
> ```

I would simplify these regexes a bit. They don't really need to enforce any syntax on the Id's (because whatever someone manages to use in a YAML "key:" to name a point/connection is legal), and having the separators outside the expressions makes them easier to understand (and eliminates the additional substr() complexity after matching):

`^points[.](.*)[.](forward|reverse)$`
`^connections[.](.*)[.](start|end)[.](ref|[0-9]*)[.](forward|reverse)$`

(Also, that last `+` in `((ref|\\d+)\\.)+` in the original kConnectionsPattern seems wrong.)

---

*[automotive/maliput/multilane/loader.cc, line 252 at r5](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF-URXj8zloIeXBGQrI:-LF-URXj8zloIeXBGQrJ:b-j9jfgs) ([raw file](https://github.com/robotlocomotion/drake/blob/ce1efe03236c3eff1df8b5b54c0ead8461b4e3b3/automotive/maliput/multilane/loader.cc#L252)):*
> ```cpp
>     parsed_reference.end = {ResolveEnd(pieces_match[3].str())};
>     // TODO(agalbachicar)    Provide support for "lane.NB" so as to reference
>     //                       the lane ID.
> ```

FYI:  I believe a lane number in a connection reference is described by just the bare integer (no `lane.` prefix).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LF1Ot0S7audqBgOy_vN:bkkwfge)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks Matt! All comments were addressed.

+@sherm1 for platform review please.

---

Review status: 1 unresolved discussion, platform LGTM missing

---

*[automotive/maliput/multilane/loader.cc, line 237 at r5](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF-Qkkm5cznUoveVvll:-LF3cYxG66kI5XKDyskJ:b-f2i1oe) ([raw file](https://github.com/robotlocomotion/drake/blob/ce1efe03236c3eff1df8b5b54c0ead8461b4e3b3/automotive/maliput/multilane/loader.cc#L237)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

I would simplify these regexes a bit. They don't really need to enforce any syntax on the Id's (because whatever someone manages to use in a YAML "key:" to name a point/connection is legal), and having the separators outside the expressions makes them easier to understand (and eliminates the additional substr() complexity after matching):

`^points[.](.*)[.](forward|reverse)$`
`^connections[.](.*)[.](start|end)[.](ref|[0-9]*)[.](forward|reverse)$`

(Also, that last `+` in `((ref|\\d+)\\.)+` in the original kConnectionsPattern seems wrong.)

</blockquote></details>

Done.

Thanks for the pointers.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LF3fhRrCzmrkZyqLuAj:b-45q3i9)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Checkpoint (still need to review tests).

---

Reviewed 1 of 10 files at r1, 1 of 8 files at r2, 2 of 8 files at r4, 2 of 5 files at r5, 1 of 1 files at r6.
Review status: 4 unresolved discussions, LGTM missing from assignee sherm1, platform LGTM missing

---

*[automotive/maliput/multilane/connection.cc, line 160 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4s1R92uufLurtze0y:-LF4s1R92uufLurtze0z:bakjhnl) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/connection.cc#L160)):*
> ```cpp
>   // coordinate and t_i the arc-length xy projection for lane_index lane.
>   const double theta_dot = type_ == kLine ? start_.z().theta_dot().value()
>                                           : start_.z().theta_dot().value() *
> ```

FYI you can also write `*start_.z().theta_dot()` rather than `start_.z().theta_dot().value()` (easier to read IMO).

---

*[automotive/maliput/multilane/loader.cc, line 170 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4tO4k9xdwmkIZ8L8b:-LF4tO4k9xdwmkIZ8L8c:b-jt5yrs) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/loader.cc#L170)):*
> ```cpp
> // Looks for the Endpoint in `point_catalog` given `endpoint_key` description.
> // `endpoint_key` should be a bare Endpoint in "points" YAML map. Either
> // ".forward" or ".reverse" keyword must be present.
> ```

Minor: is it possible to DRAKE_DEMAND (or at least ASSERT) these prerequisites? (also below).

---

*[automotive/maliput/multilane/loader.cc, line 261 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4uTGv9FRlDGLgM289:-LF4uTGv9FRlDGLgM28A:b-7fw9hf) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/loader.cc#L261)):*
> ```cpp
> // looked for inside `point_catalog` or a Connection will be selected from
> // `connection_catalog`.
> // @return An optional<StartRefernece::Spec> with the point information when
> ```

Nit: Refernece -> Reference

---

*[automotive/maliput/multilane/loader.cc, line 309 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4ugZC71KBzC50c_-K:-LF4ugZC71KBzC50c_-L:b-pfzp2n) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/loader.cc#L309)):*
> ```cpp
> // either a point in `point_catalog` or to an EndpointZ of any Connection in
> // `connection_catalog`.
> // @return An optional<EndRefernece::Spec> with the point information when
> ```

Nit: copy/paste typo

---

*[automotive/maliput/multilane/loader.cc, line 476 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4vgw68r3szpyDVfFM:-LF4vgw72X04xy69mrs_:b20gqr) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/loader.cc#L476)):*
> ```cpp
>   ParseEndpointsFromPoints(mmb["points"], &point_catalog);
> 
>   drake::log()->debug("loading raw connections !");
> ```

Minor (FYI): consider using a macro SPDLOG_DEBUG or DRAKE_SPDLOG_MACRO so that these log lines evaporate when not debugging. They are runtime checks as written (probably doesn't matter here but does sometimes).

---

*[automotive/maliput/multilane/test_utilities/multilane_types_compare.cc, line 100 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4wJIH-KSeLfSZlRYK:-LF4wJIH-KSeLfSZlRYL:b-ghx887) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/test_utilities/multilane_types_compare.cc#L100)):*
> ```cpp
>           error_message + "EndpointZ are different at theta_dot. " +
>           "z1.theta_dot(): " + std::to_string(z1.theta_dot().value()) +
>           " vs. z2.theta_dot(): " + std::to_string(z2.theta_dot().value()) +
> ```

FYI can use *theta_dot() rather than theta_dot().value().

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LF4wW53-pLgUznsvUqx:b-hvooka)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 4 unresolved discussions, LGTM missing from assignee sherm1, platform LGTM missing

---

*[automotive/maliput/multilane/connection.cc, line 160 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4s1R92uufLurtze0y:-LF94U9n7KMM8xnyyn8h:bd7zmws) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/connection.cc#L160)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Note, however, that `value()` ensures that the `optional` contains a value, but `operator*()` does not (and has undefined behavior when operating on an empty `optional`), so the two access modes are not necessarily interchangeable.
</blockquote></details>

TIL -- thanks, @maddog-tri !

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending a few minor comments.

---

Reviewed 1 of 8 files at r4, 2 of 5 files at r5.
Review status: 5 unresolved discussions, platform LGTM from [sherm1]

---

*[automotive/maliput/multilane/test/multilane_builder_test.cc, line 656 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF9OCs62lDhd5Deku6e:-LF9OCs62lDhd5Deku6f:bu7zql7) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/test/multilane_builder_test.cc#L656)):*
> ```cpp
> 
> // Holds common properties to create a Builder and Connections whose endpoint
> // information lacks of theta_dot so it is adjusted.
> ```

FYI lacks ~~of~~ theta_dot reads better

---

*[automotive/maliput/multilane/test/multilane_builder_test.cc, line 670 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF9OU4K6t5i8v8bfasV:-LF9OU4K6t5i8v8bfasW:bzckcrh) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/test/multilane_builder_test.cc#L670)):*
> ```cpp
> // theta_dot = L * sin(-atan(z_dot))
> //
> // theta_dot constants in tests below where computed in Octave using previous
> ```

Nit: where -> were

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LF9QNs479ZThLg-9pBH:bkry7ar)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks! All comments were addressed, PTAL.

---

Review status: 1 unresolved discussion, platform LGTM from [sherm1]

---

*[automotive/maliput/multilane/connection.cc, line 160 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4s1R92uufLurtze0y:-LFI6XrkAfMqK6xQtezy:b5w8t07) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/connection.cc#L160)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

TIL -- thanks, @maddog-tri !
</blockquote></details>

Moved to use `operator*()` as it has `DRAKE_DEMAND()` or checks with `has_value()` in code.

---

*[automotive/maliput/multilane/loader.cc, line 170 at r6](https://reviewable.io/reviews/robotlocomotion/drake/8676#-LF4tO4k9xdwmkIZ8L8b:-LFI186VAUkiUJY75dCi:b9s5a58) ([raw file](https://github.com/robotlocomotion/drake/blob/f609f49d3d8d5065b94131215a2db6555d43bec8/automotive/maliput/multilane/loader.cc#L170)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Minor: is it possible to DRAKE_DEMAND (or at least ASSERT) these prerequisites? (also below).
</blockquote></details>

My bad, it was a left over after refactoring code. I removed the prerequisite as it is checked `ResolveDirection`.

PTAL to `ResolveDirection` and `ResolveEnd`, I added the prerequisites to the docstrings there.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LFI8VxK4XknXyslhRlk:bma5x1z)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 5 of 5 files at r7.
Review status: all discussions resolved, platform LGTM from [sherm1], commit history needs curation

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Thanks again! I'll wait for the CI to be all green and then proceed merging this PR.

---

Review status: :shipit: complete! all discussions resolved, platform LGTM from [sherm1]

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8676#-:-LFJSmxQ6FMZ0XFQRZiz:b2tkidr)*
<!-- Sent from Reviewable.io -->
